packages/contracts/contracts/MVM/MVM_InboxSenderManager.sol

```diff
@@ -0,0 +1,111 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.9;
+
+import {Lib_AddressResolver} from "../libraries/resolver/Lib_AddressResolver.sol";
+import {iMVM_InboxSenderManager} from "./iMVM_InboxSenderManager.sol";
+
+/* Library Imports */
+
+contract MVM_InboxSenderManager is iMVM_InboxSenderManager, Lib_AddressResolver {
+    /*************
+     * Constants *
+     *************/
+    string public constant CONFIG_OWNER_KEY = "METIS_MANAGER";
+
+    /*************
+     * Variables *
+     *************/
+    // blockNumber => InboxSenderType => inboxSender
+    mapping(uint256 => mapping(InboxSenderType => InboxSender)) public inboxSenders;
+
+    uint256[] public blockNumbers;
+    mapping(InboxSenderType => address) public defaultInboxSender;
+
+    /***************
+     * Constructor *
+     ***************/
+    constructor(address _libAddressManager, InboxSender[] memory _defaultInboxSenders)
+        Lib_AddressResolver(_libAddressManager)
+    {
+        for (uint256 i = 0; i < _defaultInboxSenders.length; ++i) {
+            defaultInboxSender[_defaultInboxSenders[i].senderType] = _defaultInboxSenders[i].sender;
+        }
+    }
+
+    /**********************
+     * Function Modifiers *
+     **********************/
+    modifier onlyManager() {
+        require(
+            msg.sender == resolve("METIS_MANAGER"),
+            "MVM_InboxSenderManager: Function can only be called by the METIS_MANAGER."
+        );
+        _;
+    }
+
+    /********************
+     * Public Functions *
+     ********************/
+    function setInboxSenders(uint256 blockNumber, InboxSender[] calldata _inboxSenders)
+        external
+        override
+        onlyManager
+    {
+        _setInboxSenders(blockNumber, _inboxSenders);
+    }
+
+    // allow us to overwrite the last block number and its senders, just in case if we made any mistake
+    function overwriteLastInboxSenders(uint256 blockNumber, InboxSender[] calldata _inboxSenders)
+        external
+        override
+        onlyManager
+    {
+        require(blockNumbers.length > 0, "MVM_InboxSenderManager: No block to update.");
+
+        // pop the last block
+        uint256 lastBlockNumber = blockNumbers[blockNumbers.length - 1];
```

Comment:
It confuses me.

Why does you have a blockNumber as param, but you remove the last item of the blockNumbers?

Reply:
> It confuses me.
> 
> Why do you have a blockNumber as param, but you remove the last item of the blockNumbers?

The context of this implementation is this: @aiden069 and I were discussing a problem that if we set the wrong values of the inbox sender (e.g. we want to set 0x123...789 as inbox sender on block 10000, but the input is 0x123...789 & 100000, mis-input one more 0 of the block number),  so we would like to be able to overwrite the inbox senders of the last block.

So the logic of the implementation is this: 
1. when overwrite, we remove the last saved block and corresponding inbox senders
2. after the removal then we set the overwrite ones
3. since the remove and the add always succeed or fail at the same time, so we can guarantee the atomicity of the overwrite operation